lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -2779,56 +2782,75 @@ bool ReplaceOpenCLBuiltinPass::replaceAtomicCompareExchange(Function &F) {
+      // Extend the input to 32-bits and perform a clz/ctz.
       auto zext = builder.CreateZExt(in, ty);
-      auto clz = builder.CreateCall(clz_32bit, {zext});
-      Constant *sub_const = builder.getInt32(32 - bitwidth);
-      if (auto vec_ty = dyn_cast<VectorType>(ty)) {
-        sub_const =
-            ConstantVector::getSplat(vec_ty->getElementCount(), sub_const);
+      auto call = builder.CreateCall(func_32bit, {zext});
+      Value *tmp = call;
+      if (leading) {
+        // Clz is implemented as 31 - FindUMsb(|zext|), so adjust the result
+        // the right bitwidth.
+        Constant *sub_const = builder.getInt32(32 - bitwidth);
+        if (auto vec_ty = dyn_cast<VectorType>(ty)) {
+          sub_const =
+              ConstantVector::getSplat(vec_ty->getElementCount(), sub_const);
+        }
+        // Truncate the intermediate result to the right size.
+        tmp = builder.CreateSub(call, sub_const);
+      } else {
+        Constant *c_width = builder.getInt32(bitwidth);
+        if (auto vec_ty = dyn_cast<VectorType>(ty)) {
+          c_width =
+              ConstantVector::getSplat(vec_ty->getElementCount(), c_width);
+        }
+        auto cmp = builder.CreateICmpEQ(call, c32);
+        tmp = builder.CreateSelect(cmp, c_width, call);
       }
-      auto sub = builder.CreateSub(clz, sub_const);
-      return builder.CreateTrunc(sub, Call->getType());
+      return builder.CreateTrunc(tmp, Call->getType());
     } else {
-      // Split the input into top and bottom parts and perform clz on both. If
-      // the most significant 1 is in the upper 32-bits, return the top result
-      // directly. Otherwise return 32 + the bottom result to adjust for the
-      // correct size.
+      // Perform a 32-bit version of clz/ctz on each half of the 64-bit input.
       auto lshr = builder.CreateLShr(in, 32);
       auto top_bits = builder.CreateTrunc(lshr, ty);
       auto bot_bits = builder.CreateTrunc(in, ty);
-      auto top_clz = builder.CreateCall(clz_32bit, {top_bits});
-      auto bot_clz = builder.CreateCall(clz_32bit, {bot_bits});
-      Constant *c32 = builder.getInt32(32);
-      if (auto vec_ty = dyn_cast<VectorType>(ty)) {
-        c32 = ConstantVector::getSplat(vec_ty->getElementCount(), c32);
+      auto top_func = builder.CreateCall(func_32bit, {top_bits});
+      auto bot_func = builder.CreateCall(func_32bit, {bot_bits});
+      Value *tmp = nullptr;
+      if (leading) {
+        // For clz, if clz(top) is 32, return 32 + clz(bot).
+        auto cmp = builder.CreateICmpEQ(top_func, c32);
+        auto adjust = builder.CreateAdd(bot_func, c32);
+        tmp = builder.CreateSelect(cmp, adjust, top_func);
+      } else {
+        // For ctz, if clz(bot) is 32, return 32 + ctz(top)
+        auto bot_cmp = builder.CreateICmpEQ(bot_func, c32);
+        auto adjust = builder.CreateAdd(top_func, c32);
+        tmp = builder.CreateSelect(bot_cmp, adjust, bot_func);
```

Comment:
Could avoid the `icmp` and `select` by instead `or`ing the extended value with `(1 << bitwidth)` before calling the 32-bit ctz.

Reply:
Good idea, making that change.